### DISCUSSION TITLE
Recommended schema casing conventions

### DISCUSSION THREAD

**Discussion by sjcobb** [perses/perses#1243] [Design Discussions]
We've discussed this on calls and have been somewhat consistent so far, but I would like to propose standardizing on our casing conventions. Since we are getting closer to an initial release, I'd also like to get in breaking changes now based on feedback we've been given already. 

And since plugin developers have control over how they define their spec, these are intended to be recommendations for our core plugins as we get more contributions.

## Summary:
- **EDIT: [BREAKINGCHANGE] use camelCase in all keys**
  - Follow up to Antoine [comment](https://github.com/perses/perses/discussions/1243#discussioncomment-6316495), to align with Kubernetes [naming conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#naming-conventions) I propose changing to camelCase in all keys, see docs: "_Go field names must be PascalCase. JSON field names must be camelCase. Other than capitalization of the initial letter, the two should almost always match. No underscores or dashes in either._"
  - this proposal would be a major [BREAKINGCHANGE], particularly in the Variable spec (see [example PR](https://github.com/perses/perses/pull/1277))
- ~~Use snake_case in all keys~~
  - ~~This has been our approach so far, I'm fine with continuing with this since some of our validation relies on underscores being present, but it is worth noting that Kubernetes recommends camelCase ([ex](https://kubernetes.io/docs/reference/kubernetes-api/common-parameters/common-parameters/#propagationPolicy))~~
    - ~~We should document this in [dashboard.md](https://github.com/perses/perses/blob/main/docs/dashboard.md) docs that we've intentionally chosen to use snake_case ([related ex](https://github.com/perses/perses/discussions/1186#discussioncomment-6049892): We will use api_version or schema_version instead of apiVersion)~~
- **When `kind` is the key, use PascalCase for the value**
  - We do this already to be consistent with Kubernetes resources and it is documented in dashboard.md, not proposing any changes here
  - Related rule: whenever `kind` is used, there must be a `spec` as its sibling for consistency and to make validation easier
- **[BREAKINGCHANGE] Use kebab-case inside all other values within a plugin spec that aren't paired with `kind`**
  - _Proposing we change to this_
  - I previously suggested PascalCase everywhere since `kind` values are PascalCase, but we've had feedback that this is counter-intuitive, for example: `legend.position: 'Bottom'` looks very odd and this leaks into our component props for embed use cases, which is confusing for JavaScript / TypeScript developers
    - Kubernetes also recommends kebab-case for these values: https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/
    - Example of breaking changes, within `calculation` [schema](https://github.com/perses/perses/blob/main/schemas/common/calculation.cue): Sum -> sum, First -> first, LastNumber -> last-number
     - Within TimeSeriesChart [legend schema](https://github.com/perses/perses/blob/main/schemas/panels/time-series/time-series.cue), position: 'Bottom' -> 'bottom', mode: 'List' -> 'list', 'Table' -> 'table'

**Comment by sjcobb** [comment]
Put up a draft PR to show the proposed schema changes: https://github.com/perses/perses/pull/1262

**Comment by Gladorme** [comment]
On my side, I'm fine with this naming convention.

**Comment by AntoineThebaud** [comment]
Agree with suggested change.
Besides regarding _Kubernetes recommends camelCase_, since we plan to also support dashboards & other objects as CRDs, I think this would make sense to follow this recommendation? This would make the datamodel more coherent between 'classic' option & CRD.

**Reply by sjcobb** [reply]
I’d prefer camelCase as well, it would nice to be able to point to those recommendations in our docs and less back & forth in PRs since most FE devs are used to camelCase. It also would make the mapping in our panels simpler since the panel spec would more closely match the lower level component props.

Main reason for snake_case that Augustin has mentioned before on calls is there is some validation that relies on the underscore being there. Do you know if that is still the case / if it is, could you update that BE code? I can put up a PR soon for everything else to start using camelCase

**Reply by sjcobb** [reply]
Opened a first PR #1277 just for the variables spec changes to camelCase, I have one question about this yaml line change in list.go: https://github.com/perses/perses/pull/1277#discussion_r1253841325

**Reply by AntoineThebaud** [reply]
Thanks @sjcobb, still given this is going to produce breaking changes in many parts of the model, that was defined quite some time ago, I'd rather wait for @Nexucis's return to know what he thinks about this. I don't see a reason to reject this switch to camelCase but maybe I'm missing some context about past decisions here.

**Reply by sjcobb** [reply]
Related conversation came up in last week’s sync about `kind` and `spec` (see [notes](https://github.com/perses/perses/discussions/947?sort=new#discussioncomment-6406568)). At what levels of the schema do we need to be compatible with the Kubernetes recommendations? 

I’m still for aiming for consistency everywhere, but it was pointed out that inside the definition, we have more freedom.

**Reply by Nexucis** [reply]
> Main reason for snake_case that Augustin has mentioned before on calls is there is some validation that relies on the underscore being there. Do you know if that is still the case / if it is, could you update that BE code? I can put up a PR soon for everything else to start using camelCase

The initial convention was coming from the BE config that is following the `snake_case` convention. Because indeed we are basing the mapping between environment variable and the config on the `_`.

But this part is completely isolated and doesn't impact the data model.
I am totally fine with the convention chosen and it is a good thing we have defined it !

> At what levels of the schema do we need to be compatible with the Kubernetes recommendations?

At the first level ie `kind, metadata, spec`.
So at plugin level we are completely free to do whatever we want.

I think it remains some last few changes to be done. And after that we should be able to close this conversation :)

**Comment by AntoineThebaud** [comment]
Finished the implementation of the new convention with https://github.com/perses/perses/pull/1410.
